File: leetcode/tree/235. Lowest Common Ancestor of a Binary Search Tree/py/main.py
class Solution:
    # Time O(N)
    # Space O(N)
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode',
                             q: 'TreeNode') -> 'TreeNode':
        def dfs(node):
            if node is None:
                return None

            if node.val > p.val and node.val > q.val:
                return dfs(node.left)

            if node.val < p.val and node.val < q.val:
                return dfs(node.right)

            return node

        return dfs(root)

File: leetcode/tree/235. Lowest Common Ancestor of a Binary Search Tree/js/main.js
// Можно решить так же как 236 задачу

// Однако лучше воспользоваться свойством бинарного дерева поиска
// Time O(N)
// Space O(N)
const lowestCommonAncestor = (root, p, q) => {
  if (p.val > root.val && q.val > root.val) {
    return lowestCommonAncestor(root.right, p, q);
  }

  if (p.val < root.val && q.val < root.val) {
    return lowestCommonAncestor(root.left, p, q);
  }

  return root;
};

// Time O(N)
// Space O(1)
const lowestCommonAncestor2 = (root, p, q) => {
  while (root !== null) {
    if (p.val > root.val && q.val > root.val) {
      root = root.right;
    } else if (p.val < root.val && q.val < root.val) {
      root = root.left;
    } else {
      return root;
    }
  }
};
